fuel-core/src/schema/contract.rs

```diff
@@ -113,4 +121,112 @@ impl ContractBalanceQuery {
             asset_id,
         })
     }
+
+    async fn contract_balances(
+        &self,
+        ctx: &Context<'_>,
+        filter: ContractBalanceFilterInput,
+        first: Option<i32>,
+        after: Option<String>,
+        last: Option<i32>,
+        before: Option<String>,
+    ) -> async_graphql::Result<Connection<AssetId, ContractBalance, EmptyFields, EmptyFields>> {
+        let db = ctx.data_unchecked::<Database>().clone();
+
+        let balances: Vec<ContractBalance> = db
```

Comment:
this should be moved inside of the query section below, where the start asset and number of results to fetch are specified. 

Otherwise, if we collect the iterator into a vec now, and then filter it based on the query this is very inefficient. 

For example, let's say I have an exchange contract with 1000 asset balances. Even if my application only showed 5 balances per page, collecting the iterator right will cause all 1000 to load into memory, and then discard 995 of the results below. 

We should use the before/after cursors to tell the database where to start the iterator, to avoid loading all these extra values into memory. Then we should only consume as many items from the database iterator as the user requested, so that we don't fetch anything past that point.

Move this into the query section below, and populate the two `None` values with the query params. Only collect once the iterator is configured to touch the least amount of state.

Reply:
We should update that in a separate PR. Coin balances is more complicated, because the cursor is across asset id's, but to compute a balance it needs to sum up all the coins for each asset. We need to be careful to ensure we don't accidentally compute partial balances by applying the paging params like first/last directly to the coins iterator.

Reply:
Ok, will resolve this one first and make that a seperate issue then